scheduler/scripts/run_main.py

```diff
@@ -147,39 +152,68 @@ def main(*,
+                    end_timeslot_bounds = {site: TimeslotIndex(event_start_time_slot)}
+
+                # If tbe following conditions are met:
+                # 1. there are plans (i.e. plans have been generated by this point, meaning at least the
+                #    evening twilight event has occurred); and
+                # 2. a new plan is to be produced (TODO: GSCHED-515)
+                # then perform time accounting.
+                #
+                # This will also perform the final time accounting when the night is done and the morning twilight
+                # event has occurred.
+                if plans is not None:
+                    logger.info(f'Performing time accounting for night index {night_idx} '
+                                f'at {site.name} up to timeslot {event_start_time_slot}.')
+                    collector.time_accounting(plans,
+                                              sites=frozenset({site}),
+                                              end_timeslot_bounds=end_timeslot_bounds)
+
+                # If the following conditions are met:
+                # 1. the night is not done;
+                # 2. a new plan is to be produced (TODO: GSCHED-515)
+                # fetch a new selection and produce a new plan.
+                if not night_done:
+                    logger.info(f'Retrieving selection for night index {night_idx} '
+                                f'at {site.name} starting at time slot {event_start_time_slot}.')
                     selection = selector.select(night_indices=night_indices,
                                                 sites=frozenset([event.site]),
                                                 starting_time_slots={site: {night_idx: event_start_time_slot
                                                                             for night_idx in night_indices}})
 
-                    # Run the optimizer to get the plans for the first night in the selection.
-                    plans = optimizer.schedule(selection)
-                    night_timeline.add(NightIndex(night_idx),
-                                       site,
-                                       TimeslotIndex(event_start_time_slot),
-                                       event,
-                                       plans[0][site])
-
-                # Perform the time accounting. If the night is done, we execute it to completion.
-                end_timeslot_bounds = None if night_done else {site: TimeslotIndex(event_start_time_slot)}
-                collector.time_accounting(plans[0],
-                                          sites=frozenset({site}),
-                                          end_timeslot_bounds=end_timeslot_bounds)
+                    # Right now the optimizer generates List[Plans], a list of plans indexed by
+                    # every night in the selection. We only want the first one, which corresponds
```

Comment:
This look like a hack, we should change that to one 'Plans' or Site Plans.

Reply:
What GM should be doing is creating one plan per site for the given night, although right now, it creates a `List[Plan]` for all the nights for which it has data in the `Selection`. It's a bit hacky, but it's flexible: I just make sure that we only have one night in the `Selection` and GM then only prepares a `List[Plans]` with that one night.

Ultimately, I'm not sure what we are going to want: GM to still be able to produce `Plans` across multiple nights, or simplify it down to one `Plans` for the current night. The latter will, I think, require a lot of propagation through the code and offer less generalized functionality, but until those decisions are made, this is going to remain a bit hacky but functional.